python/ray/data/dataset.py

```diff
@@ -1354,17 +1355,52 @@ def to_spark(self,
         core_worker = ray.worker.global_worker.core_worker
         locations = [
             core_worker.get_owner_address(block)
-            for block in self.get_blocks()
+            for block in self.get_internal_block_refs()
         ]
         return raydp.spark.ray_dataset_to_spark_dataframe(
-            spark, self.schema(), self.get_blocks(), locations)
+            spark, self.schema(), self.get_internal_block_refs(), locations)
 
-    def to_pandas(self) -> List[ObjectRef["pandas.DataFrame"]]:
+    def to_pandas(self, limit: int = 1000) -> "pandas.DataFrame":
+        """Convert this dataset into a single Pandas DataFrame, limiting the
+        number of records returned.
+
+        This is only supported for datasets convertible to Arrow records.
+
+        Time complexity: O(limit)
+
+        Args:
+            limit: The maximum number of records to return.
+
+        Returns:
+            A Pandas DataFrame created from this dataset, containing a limited
+            number of records.
+        """
+
+        batcher = Batcher(batch_size=limit)
+
+        for idx, block in enumerate(self._blocks):
+            block = ray.get(block)
```

Comment:
Won't this fetch excess data if the first block is much bigger than the limit? How about calling .limit() and then gathering all the blocks?